src/getPublicKeys.c

```diff
@@ -0,0 +1,265 @@
+		ASSERT(ctx->responseReadyMagic == RESPONSE_READY_MAGIC);
+
+		io_send_buf(SUCCESS, (uint8_t*) &ctx->extPubKey, SIZEOF(ctx->extPubKey));
+		ctx->responseReadyMagic = 0; // just for safety
+
+		if (ctx->remainingKeys > 0) {
+			ui_displayBusy(); // waiting for another APDU
+		} else {
+			ui_idle(); // we are done, display the main app screen
+		}
+	}
+	UI_STEP_END(GET_KEY_UI_STEP_INVALID);
+}
+
+// derive the key described by ctx->pathSpec and run the ui state machine accordingly
+// if isSingleKeyExport is true, more restrictive security policy is applied
+static void returnOneKey(bool isSingleKeyExport)
+{
+	ctx->responseReadyMagic = 0;
+
+	// Check security policy
+	security_policy_t policy = (isSingleKeyExport) ? 
+			policyForGetExtendedPublicKey(&ctx->pathSpec) :
+			policyForGetExtendedPublicKeyBulkExport(&ctx->pathSpec);
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+
+    {
+        // Calculation
+        deriveExtendedPublicKey(
+                & ctx->pathSpec,
+                & ctx->extPubKey
+        );
+        ctx->responseReadyMagic = RESPONSE_READY_MAGIC;
+    }
+
+	switch (policy) {
+#	define  CASE(policy, step) case policy: {ctx->ui_step = step; break;}
+		CASE(POLICY_PROMPT_WARN_UNUSUAL,    GET_KEY_UI_STEP_WARNING);
+		CASE(POLICY_PROMPT_BEFORE_RESPONSE, GET_KEY_UI_STEP_DISPLAY);
+		CASE(POLICY_ALLOW_WITHOUT_PROMPT,   GET_KEY_UI_STEP_RESPOND);
+#	undef   CASE
+	default:
+		ASSERT(false);
+	}
+	getPublicKeys_getOneKey_ui_runStep();
+}
+
+// ============================== INIT ==============================
+
+enum {
+	HANDLE_INIT_UI_STEP_CONFIRM = 100,
+	HANDLE_INIT_UI_STEP_RESPOND,
+	HANDLE_INIT_UI_STEP_INVALID,
+} ;
+
+static void getPublicKeys_handleInit_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	ui_callback_fn_t* this_fn = getPublicKeys_handleInit_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step);
+	UI_STEP(HANDLE_INIT_UI_STEP_CONFIRM) {
+        char secondLine[100];
+        explicit_bzero(secondLine, SIZEOF(secondLine));
+        snprintf(secondLine, SIZEOF(secondLine), "%u public keys?", ctx->remainingKeys);
+        ASSERT(strlen(secondLine) < SIZEOF(secondLine));
+
+		ui_displayPrompt(
+		        "Confirm export",
+		        secondLine,
+		        this_fn,
+		        respond_with_user_reject
+		);
+	}
+	UI_STEP(HANDLE_INIT_UI_STEP_RESPOND) {
+		ASSERT(ctx->remainingKeys > 0);
+		returnOneKey(false); // runs another UI state machine
+
+		ctx->stage = GET_KEYS_STAGE_GET_KEYS;
+	}
+	UI_STEP_END(HANDLE_INIT_UI_STEP_INVALID);
+}
+
+static void getPublicKeys_handleInitAPDU(uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		CHECK_STAGE(GET_KEYS_STAGE_INIT);
+
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		// parse data
+
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		parsePath(&view);
+
+		size_t remaining = view_remainingSize(&view);
+		switch (remaining) {
```

Comment:
what's the point of making this backwards compatible? We own both JS and device and so JS can be responsible for proper interaction